README.md
AdaptaSlider
============

A simple/basic/bare-bones slider/carousel jQuery plugin that gives you the functionality you need without requiring you to change your design or code.

You've just finished with your meticulous HTML markup and CSS styling for the main page of a new site. Designed the slider to look a certain way, but now you need to make it functional. If you go searching for a script, it's likely you'll have to restructure the way you made your slider and change a bunch of CSS to jive with the script's expected setup... Or, you can try to build a slider from the ground up, which you could do, but shouldn't there be something out there already that isn't so restrictive?

Every time I need to implement a slider I go in search of a simple, adaptable slider script that doesn't impose it's structure or style on what I've already designed. I always fail and settle for hacking my code so that it fits what the slider script expects, and I usually resort to editing the script anyways.

Why should a slider script insist on how you make your navigation buttons? Why do some insist that the _script_ should create the buttons without asking _you_ what they should look like? Why do I have to include an extra CSS file? It's as if they expect that you choose a slider script _before_ designing your site...

This time I decided to search no longer and instead build what I was looking for myself. Enter **AdaptaSlider**. While I don't believe that I can personally come up with The Solution, I think this is close.

Why another jQuery slider?
--------------------------

With the [ubiquity of jQuery](http://w3techs.com/technologies/details/js-jquery/all/all), it's likely that you're already using it. WordPress, Drupal, Joomla, and many other CMS softwares include jQuery, and Twitter Bootstrap relies on it for extra functionality. And let's be honest: jQuery makes our lives easier, especially concerning cross-browser compatibility. **AdaptaSlider** was built with jQuery 1.11.0 so that it should work in IE 6, 7, & 8.

What's different about AdaptaSlider?
------------------------------------

**AdaptaSlider** is specifically designed to be used within an existing design, meaning that you have total control over the style of the slider, the slides, the navigation buttons... everything! On the flip side, it means that you _must_ design the slider, the slides, and the navigation buttons yourself. I've included some sample CSS for each example page, but it's your responsibility to make the slider unique for your instance.

**AdaptaSlider** has a few tiers of functionality available, depending on your needs.

### 1. Skimpy

You can use **AdaptaSlider** as a basic timer and class-setter with optional navigation functionality. This means that it will not apply any CSS overrides to your HTML, and will not control any transitions; all it does is set `current`, `prev`, and `next` class names on the slide elements (and `active` on the select buttons, if provided). And these class names are customizable!

This means that you have total control, using CSS transitions or even JavaScript/jQuery, to create the transitions yourself.

You don't even have to use the timer; you can tell it what elements to use as navigation controls, or you can forego buttons altogether if you want to call the `next()`, `prev()`, or `select()` methods yourself with JS/jQ.

### 2. Scanty

Building on the last tier, the middle-of-the-road option lets you do all that _with_ automatic transitions. These transitions are not fancy, but you are welcome to override the `transition(fromSlide, toSlide)` method for more exciting animations. Otherwise, your choices are `slide` (default) and `fade`.

This is how **AdaptaSlider** is set up by default. When you apply the plugin to an element, it's children will be arranged in the appropriate way and the slider will autoplay.

### 3. Plenty

There are plenty of ways to customize **AdaptaSlider** to fit your needs.

As mentioned above, you can decide whether the slider animates, autoplays, has buttons, and what the buttons are.

There are also a bunch of other options for many different configurations. Advanced users can specify callback functions to be triggered during specific parts of the slider lifecycle, as well as pass custom setup & transition function in without having to edit the source code. All options are detailed in the **Settings** section below.

How do I use it?
----------------

### HTML Setup

**AdaptaSlider** is designed to be as flexible as possible, but your slider HTML should at least follow a "parent element with all children as slides" paradigm. This means that you should be able to use list elements, a `<div>` with child `<div>`s or `<img>`s, a `<section>` with `<article>`s, etc. However, I would generally advise using an unordered list:

```html
<ul id="slider" role="listbox">
	<li><img src="example/1.jpg" /></li>
	<li><img src="example/2.jpg" /></li>
	<li><img src="example/3.jpg" /></li>
	<li><img src="example/4.jpg" /></li>
</ul>
```

### Default Use

To use **AdaptaSlider** with the above defaults (sliding transitions on a 7 second timer, no controls) simply load jQuery in the `<head>`:

```html
<head>
	...
	<script src="http://ajax.googleapis.com/ajax/libs/jquery/1.10.2/jquery.min.js"></script>
</head>
```

Then load **AdaptaSlider** and call it on a jQuery object at the end of `<body>`:

```html
<body>
	...
	<script src="js/jquery.AdaptaSlider.js"></script>
	<script>
		// $(document).ready equivalent
		$(function() {
			$('#slider').AdaptaSlider();
		});
	</script>
</body>
```

### Basic Use

The whole point of this is that you're probably going to want some customization. In that case, call the plugin using the options described above as properties in an object passed as a parameter to `AdaptaSlider()`:

```javascript
$('#slider').AdaptaSlider({
	mode: 'fade',
	easing: 'linear',
	autoPlay: false,
	transitionTime: 1000
});
```

#### Navigation

The most important customizeable feature is the way navigation controls are handled. You have the option of specifying existing elements, having the plugin create them, or no navigation (default).

##### Specifying existing navigation elements

If you have already created and styled slider navigation controls to your liking, you can tell **AdaptaSlider** to use them!

```javascript
$('#slider').AdaptaSlider({
	prevButton: '.prev-button-class',
	nextButton: '.next-button-class',
	selectButton: '.select-button-class'
});
```

This will provide functionality to your prev and next buttons and will look for any elements with `.select-button-class` to associate with slides in document order.

##### Creating navigation elements

You can also have **AdaptaSlider** programatically create controls for you with the `prevHTML`, `nextHTML`, and `selectHTML` settings. These are simply inputs for HTML strings (opening & closing tags, please) that the script should insert into the element specified by the respective `*Button` setting.

```javascript
$('#slider').AdaptaSlider({
	prevButton: '.prev-button-container',
	prevHTML: '<button class="prev">prev</button>',
	nextButton: '.next-button-container',
	nextHTML: '<button class="next">next</button>',
	selectButton: '.select-button-container',
	selectHTML: '<button class="select"></button>',
	selectHTMLNumbers: true
});
```

Select buttons are unique in that there will be one created for each slide. You can also use the `selectHTMLNumbers` option to automatically insert numbers, starting at 1.

### Advanced Use

Setting `mode` to "custom" will give you the most control over the slider transitions. Doing so limits the script to assigning `prev`, `current`, and `next` classes (which are customizable) to the correct slides at the appropriate time intervals. You can use CSS to create your own transitions:

```css
#slider {
	list-style: none;
	margin: 0;
	padding: 0;
	height: 0;
	padding-bottom: 66%;
	background-color: #333;
	position: relative;
	overflow: hidden;
}
#slider li {
	position: absolute;
	top: 0;
	left: 0;
	display: none;
	width: 100%;
	height: 100%;
	/*padding-bottom: 66%;*/
	background-repeat: no-repeat;
	background-position: center center;
	background-size: 100% auto;

	-webkit-transition-property: top, left, opacity;
	-moz-transition-property: top, left, opacity;
	-o-transition-property: top, left, opacity;
	transition-property: top, left, opacity;

	-webkit-transition-duration: 1s;
	-moz-transition-duration: 1s;
	-o-transition-duration: 1s;
	transition-duration: 1s;
}
#slider .prev {
	display: block;
	top: 100%;
	opacity: 0;
	z-index: 3;
}
#slider .current {
	display: block;
	left: 0;
	opacity: 1;
	z-index: 2;
}
#slider .next {
	display: block;
	left: 100%;
	opacity: 0;
}
```

See Custom Transitions and API sections below for more advanced usage.

Settings
--------

#### Slideshow features

- `mode : 'slide'` Either 'slide', 'fade', or 'custom'. Only the first letter is checked, will default to slide (req: animate)
- `easing: 'swing'` Either 'swing' or 'linear'. You can include the easing jQuery plugin for more options: http://gsgd.co.uk/sandbox/jquery/easing/
- `startSlide: 0` Zero-indexed slide to start on
- `animate: true` If false, slider will not control transitions; you can then create CSS transitions based on the 'current', 'prev', & 'next' classes that are applied to the slides
- `autoPlay: true` If false, slider must be manually controlled
- `pauseOnHover: true` If false, slider will not pause on hover (req: autoPlay)
- `loop: true` If false, slider will stop on last slide (req: autoPlay)

#### Custom class names

- `currentSlideClass: 'current'` Class applied to current slide
- `prevSlideClass: 'prev'` Class applied to previous slide
- `nextSlideClass: 'next'` Class applied to next slide
- `activeSelectClass: 'active'` Class applied to active select button

#### Timing

- `pauseTime : 7000` Milliseconds to pause on each slide (req: autoPlay)
- `transitionTime : 600` Milliseconds to transition between slides (req: autoPlay)
- `resumeTime: 12000` Milliseconds to resume autoplay after user interaction (req: autoPlay)

#### Controls

- `useControls: true` If false, controls will not be set up; you can still manually invoke transitions with your own script
- `prevButton: ''` jQuery selector string for prev button
- `nextButton: ''` jQuery selector string for next button
- `selectButton: ''` jQuery selector string for select button(s)

#### Button creation

- `prevHTML: ''` This string (or function returning an HTML string) will be placed inside prevButton (req: useControls, prevButton)
- `nextHTML: ''` This string (or function returning an HTML string) will be placed inside nextButton (req: useControls,nextButton)
- `selectHTML: ''` This string (or function(index) returning an HTML string) will be placed inside selectHTML, one for each slide (req: useControls, selectHTML)
- `selectHTMLNumbers : false` Whether numbers should be inserted into the selectHTML; incompatible with selectHTML containing a function (req: useControls, selectHTML)

#### Callback functions

- `onLoaded: function(){}` Callback triggered after plugin loads
- `onWillSlide: function(fromSlide, toSlide){}` Callback triggered before slider slides, supplies 'fromSlide' & toSide'
- `onDidSlide: function(fromSlide, toSlide){}` Callback triggered after slider slides, supplies 'fromSlide' & toSide'

#### Custom
- `customSetup: function(slider){}` (req: mode=custom)
- `customTransition: function(slider, fromSlide, toSlide, time, easing, callback){}` You MUST invoke the callback! (req: mode=custom)

#### Debug

- `verbose: false` If true, log all the things to JavaScript console

Custom Transitions
------------------

In order to use the `customSetup()` and `customTransition()` options, `mode` must be set to "custom". You can then pass in functions for slide setup and transition like so:

```javascript
$('#slider').adaptaSlider({
    mode: 'custom',
    customSetup: function(slider) {
        // Here you have access to the 'slider' jQuery object.
        // You can also use anything in the public API with 'this', ie:
        this.$slider.css({...});
        this.$slides.css({...});
    },
    customTransition: function(slider, fromSlide, toSlide, time, easing, callback) {
        // Here you get the slider as well as fromSlide & toSlide (int),
        // and also the time & easing as set by the options.
        // To get toSlide as a jQuery object, use it as the index in the .eq() function:
        this.$slides.eq(toSlide).animate({...}, time, easing, callback);
        // Note the 'callback' at the end of the .animate() function; this is important!
        // It notifies the plugin that the transition is finished. Don't forget it!
        // If you're not using jQuery's .animate(), just ensure that you call the callback
        // _after_ your transition is complete. You can call it manually like this:
        callback.call(this);
        // Or manually trigger this function:
        this.doneAnimating();
    }
});
```

API
---

The quickest way to get access to the **AdaptaSlider** API is to invoke it like this:

```javascript
var sliderAPI = $('#slider').adaptaSlider({...}).data('adaptaSlider');
```

(If you know of a more elegent way to provide the API while still preserving jQuery chainability, please let me know!)

You now have access to the API through the `sliderAPI` variable (or whatever you choose to name it).

### Public Properies

- `slider` The slider DOM element
- `$slider` The slider jQuery object
- `$slides` All slides as a jQuery object; Accessable after `update()`
- `numSlides` Number of slides (int); Accessable after `update()`
- `$prevButton` The prevButton jQuery object; Accessable after `update()`
- `$nextButton` The nextButton jQuery object; Accessable after `update()`
- `$selectButton` The selectButton jQuery object; Accessable after `update()`
- `settings` The settings object; see **Settings** above
- `playing` Whether auto-playing or not (bool); treat this as read-only and use `setPlaying()` or `togglePlaying()'
- `animating` True while transitioning (bool); trat this as read-only (see **Custom Transitions** section above)
- `timer` Timer ID for the current transition timer (int); treat this as read-only, use `setTimer()` and `clearTimer()`

### Public Methods

- `update()` Update the slider; invoke if you add slides or nav elements or change settings
- `prev()` Go to previous slide
- `next()` Go to next slide
- `select(index)` Go to slide at `index` (int)
- `getPrevSlide()` Returns previous slide (int)
- `getNextSlide()` Returns next slide (int)
- `setCurrentSlide(index)` Sets slide classes; does not initiate transition
- `setCurrentSelect()` Sets active select button class (based on currentSlide)
- `setCurrentSlideAndSelect(index)` Convenience method to set slide & select button classes at once
- `setTimer(time, callback)` Calls `callback` after the `time` has elapsed; this is used internally for pauses between transitions, so using this with `animate: true` will cause issues
- `clearTimer()` Clears the timer set by `setTimer()`
- `setPlaying(bool)` Turns autoplay on or off
- `togglePlaying()` Toggles autoplay
- `doneAnimating()` Used to notify plugin that the animation is finished

To Do
=====

I'm considering dropping jQuery .animate() transitions. There are some bugs, particularly on older versions of Safari, that would be eliminated if I used CSS transitions. I fear, however, that this change would require adding CSS, which I didn't want to require when I started out writing this. I may just bite the bullet and do it, as it would also provide some succinct examples of how to create your own transitions with CSS.


CHANGELOG.md
AdaptaSlider
============

Changelog
---------

1.0.0
- Initial release

js/jquery.adaptaslider.js
/*
 * AdaptaSlider v1.0.1
 * http://github.com/cr0ybot/adaptaslider
 *
 * Made by Cory Hughart
 * MIT license: http://opensource.org/licenses/MIT
 *
 * Built with jQuery 1.11.0
 */

;(function($, window, document, undefined) {
	"use strict";

	// Defaults
	var pluginName = 'adaptaSlider',
		defaults = {

			// Slideshow features
			mode                : 'slide',      // set to 'slide', 'fade', or 'custom' (only first letter checked, will default to slide) (req: animate)
			easing              : 'swing',      // 'linear' or 'swing', can include easing plugin for more options: http://gsgd.co.uk/sandbox/jquery/easing/ (req: animate)
			startSlide          : 0,            // zero-indexed slide to start on
			animate             : true,         // if false, slider will not control transitions; you can then create CSS transitions based on the 'current', 'prev', & 'next' classes that are applied to the slides
			autoPlay            : true,         // if false, slider must be manually controlled
			pauseOnHover        : true,         // if false, slider will not pause on hover (req: autoPlay)
			loop                : true,         // if false, slider will stop on last slide (req: autoPlay)

			// Custom class names (set to empty string to not add class)
			sliderClass         : 'slider',     // Class applied to slider
			currentSlideClass   : 'current',    // Class applied to current slide
			prevSlideClass      : 'prev',       // Class applied to previous slide
			nextSlideClass      : 'next',       // Class applied to next slide
			activeSelectClass   : 'active',     // Class applied to active select button

			// Timing
			pauseTime           : 7000,         // milliseconds to pause on each slide (req: autoPlay)
			transitionTime      : 600,          // milliseconds to transition between slides (req: autoPlay)
			resumeTime          : 12000,        // milliseconds to resume autoplay after user interaction (req: autoPlay)

			// Controls
			useControls         : true,         // if false, controls will not be set up; you can still manually access controls with your own script
			prevButton          : '',           // jQuery selector string for prev button; required for prevHTML
			nextButton          : '',           // jQuery selector string for next button; required for nextHTML
			selectButton        : '',           // jQuery selector string for select button(s); required for selectHTML

			// Button creation
			prevHTML            : '',           // this string (or function returning an HTML string) will be placed inside prevButton (req: useControls, prevButton)
			nextHTML            : '',           // this string (or function returning an HTML string) will be placed inside nextButton (req: useControls,nextButton)
			selectHTML          : '',           // this string (or function(index) returning an HTML string) will be placed inside selectHTML (multiplied by number of slides) (req: useControls, selectHTML)
			selectHTMLNumbers   : false,        // Whether numbers should be inserted into the selectHTML (incompatible with selectHTML containing function)

			// Callback functions
			onLoaded            : function(){},                     // Callback triggered after plugin loads
			onWillSlide         : function(fromSlide, toSlide){},   // Callback triggered before slider slides, supplies 'fromSlide' & toSide'
			onDidSlide          : function(fromSlide, toSlide){},   // Callback triggered after slider slides, supplies 'fromSlide' & toSide'

			// Custom
			customSetup         : function(slider){},               // (req: mode=custom)
			customTransition    : function(slider, fromSlide, toSlide, time, easing, callback){
				setTimeout(callback, time, this);
			}, // You MUST invoke the callback! (req: mode=custom)

			// Debug
			verbose             : false         // If true, log to JavaScript console; leave false for production
		};

	function AdaptaSlider(element, options) {

		// Setup self & settings shortcut vars
		var self = this, s;

		self.slider = element;
		self.$slider = $(element);
		// Shortcut to settings object 's'
		self.settings = s = $.extend({}, defaults, options);
		self.currentSlide = 0;
		self.playing = false;
		self.animating = false;
		self.timer = 0;

		// Debug output
		var debug = {
			log: function() {
				if (s.verbose) console.log.apply(console, arguments);
			},
			warn: function() {
				if (s.verbose) console.warn.apply(console, arguments);
			},
			dir: function() {
				if (s.verbose) console.dir.apply(console, arguments);
			},
			group: function() {
				if (s.verbose) console.group.apply(console, arguments);
			},
			groupEnd: function() {
				if (s.verbose) console.groupEnd.apply(console, arguments);
			}
		}

		debug.log('as: %s initiated', pluginName);

		// Debug
		debug.log('as: slider object:');
		debug.dir(self.slider);
		debug.log('as: settings:');
		debug.dir(s);

		// Add reverse reference to the DOM object
		// This prevent multiple instances, but also
		// provides a way to access the plugin via
		// element.data('adaptaSlider').publicMethod() or
		// element.data('adaptaSlider').settings.propertyName
		self.$slider.data(pluginName, self);

		// Object containing all animations
		self.animate = {
			// Fade transition
			fade: function(fromSlide, toSlide, time, easing, callback) {
				var f = self.$slides.eq(fromSlide);
				var t = self.$slides.eq(toSlide);
				t.css('visibility', 'visible');
				f.fadeTo(time, 0, easing, function() {
					f.css('visibility', 'hidden');
				});
				t.fadeTo(time, 1, easing, callback);
			},

			// Slide transition
			slide: function(fromSlide, toSlide, time, easing, callback) {
				var pos = -toSlide*100;
				debug.log('slide to position: %d%', pos);
				self.$slides.eq(0).animate({
					'margin-left': pos + '%'
				}, time, easing, callback);
			}
		}

		// Private constructor, only called from within
		var init = function() {
			debug.log('as: init()');

			// Setup slider classes
			self.$slider.addClass(s.sliderClass);
			//self.$slides.addClass('slide');

			// Setup initial state
			self.currentSlide = s.startSlide;
			self.update();

			// Trigger on loaded callback
			onLoadedCallback();

			return self;
		};

		// Public methods

		self.update = function() {
			debug.log('as: update()');

			// Find slides
			self.$slides = self.$slider.children();
			self.numSlides = self.$slides.length;

			// Do nothing if there's only one slide
			if (self.numSlides > 1) {
				// Create controls
				createControls();

				// Current slide & slide classes
				self.setCurrentSlideAndSelect(self.currentSlide);

				// Setup for animation
				animationSetup();

				// Start slider
				if(s.autoPlay){
					self.setPlaying(true);

					// Pause on hover
					if (s.pauseOnHover && s.pauseTime && s.pauseTime > 0) {
						self.$slider.hover(
							function () {
								if (self.playing) {
									debug.log('as: pausing for hover');
									//self.togglePlaying();
									self.clearTimer();
									self.$slider.addClass('paused');
								}
							},
							function () {
								if (self.playing) {
									debug.log('as: resuming after hover');
									//self.togglePlaying();
									self.setTimer();
									self.$slider.removeClass('paused');
								}
							}
						);
					}
				}
			}
		};

		// Prev
		self.prev = function() {
			debug.log('as: prev()');
			goToSlide(self.getPrevSlide());
		};

		// Next
		self.next = function() {
			debug.log('as: next()');
			goToSlide(self.getNextSlide());
		};

		// Select
		self.select = function(slide) {
			debug.log('as: select(%d)', slide);
			goToSlide(modInt(parseInt(slide), self.numSlides));
		};

		self.getPrevSlide = function() {
			debug.log('as: getPrevSlide()');
			return modInt(self.currentSlide - 1, self.numSlides);
		};

		self.getNextSlide = function() {
			debug.log('as: getNextSlide()');
			return modInt(self.currentSlide + 1, self.numSlides);
		};

		self.setCurrentSlide = function(slide) {
			debug.log('as: setCurrentSlide(%d)', slide);
			self.currentSlide = modInt(parseInt(slide), self.numSlides) || 0;
			self.$slides.removeClass(s.currentSlideClass+' '+s.prevSlideClass+' '+s.nextSlideClass);
			self.$slides.eq(self.currentSlide).addClass(s.currentSlideClass);
			self.$slides.eq(self.getPrevSlide()).addClass(s.prevSlideClass);
			self.$slides.eq(self.getNextSlide()).addClass(s.nextSlideClass);
		};

		self.setCurrentSelect = function() {
			debug.log('as: setCurrentSelect()');
			debug.group('changing active select button...');
			if (self.$selectButton) {
				debug.log('setting selectButton:' + self.currentSlide + ' to active');
				self.$selectButton.removeClass(s.activeSelectClass);
				self.$selectButton.eq(self.currentSlide).addClass(s.activeSelectClass);
			}
			else {
				debug.warn('no selectButtons found');
			}
			debug.groupEnd();
		};

		self.setCurrentSlideAndSelect = function(slide) {
			self.setCurrentSlide(slide);
			if (self.$selectButton) self.setCurrentSelect();
		};

		self.setTimer = function(time, callback) {
			var _time = time ? time : s.pauseTime;
			// default callback to self.next() if not provided
			var _callback = callback ? callback : self.next;
			debug.log('as: setTimer(%d, %s)', _time, (_callback.name ? _callback.name : '[anonymous]'));
			self.clearTimer();
			self.timer = setTimeout(function() { _callback.call(self); }, _time);
		};

		self.clearTimer = function() {
			debug.log('as: clearTimer()');
			clearTimeout(self.timer);
		};

		self.setPlaying = function(playing) {
			self.playing = playing;
			debug.log('as: playing set to %s', self.playing);
			if (self.playing) {
				self.setTimer();
			}
			else {
				self.clearTimer();
			}
		};

		self.togglePlaying = function() {
			self.setPlaying(!self.playing);
		};

		self.doneAnimating = function() {
			self.animating = false;
		};

		// Private methods

		// Customizable on load callback
		var onLoadedCallback = function() {
			debug.log('as: onLoadedCallback()');
			s.onLoaded.call(self);
		};

		// Customizable on will slide callback
		var onWillSlideCallback = function(fromSlide, toSlide) {
			debug.log('as: onWillSlideCallback(%d, %d)', fromSlide, toSlide);
			s.onWillSlide.call(self, fromSlide, toSlide);
		};

		// Customizable on did slide callback
		var onDidSlideCallback = function(fromSlide, toSlide) {
			debug.log('as: onDidSlideCallback(%d, %d)', fromSlide, toSlide);
			s.onDidSlide.call(self, fromSlide, toSlide);
		};

		// Handles creation of slider controls
		var createControls = function() {
			debug.log('as: createControls()');
			// Only create controls if wanted or needed
			if (s.useControls && self.numSlides > 1) {
				debug.group('creating controls...');
				// Setup prev button
				if (s.prevButton && $(s.prevButton).length) {
					debug.log('creating prevButton');

					self.$prevButton = $(s.prevButton);

					// If constructor is provided, append to prevButton element
					if (s.prevHTML) {
						if (typeof s.prevHTML == 'string') {
							debug.log('adding prevHTML');
							// Check for valid HTML
							if (validateHTML(s.prevHTML, 'prevHTML'))
								self.$prevButton = $(s.prevHTML).appendTo(self.$prevButton);
					   }
					   else if (typeof s.prevHTML == 'function') self.$prevButton = $(s.prevHTML.call(this)).appendTo(self.$prevButton);
					}

					// Click listener
					self.$prevButton.click(function(e) {
						e.preventDefault();
						self.prev();
					});
				}
				else {
					debug.warn('prevButton not specified');
				}

				// Setup next button
				if (s.nextButton && $(s.nextButton).length) {
					debug.log('creating nextButton');

					self.$nextButton = $(s.nextButton);

					// If constructor is provided, append to prevButton element
					if (s.nextHTML) {
						//self.$nextButton = self.$nextButton.append(s.nextHTML);
						if (typeof s.nextHTML == 'string') {
							debug.log('adding nextHTML');
							// Check for valid HTML
							if (validateHTML(s.nextHTML, 'nextHTML'))
								self.$nextButton = $(s.nextHTML).appendTo(self.$nextButton);
						}
						else if (typeof s.nextHTML == 'function')
							self.$nextButton = $(s.nextHTML.call(this)).appendTo(self.$nextButton);
					}

					// Click listener
					self.$nextButton.click(function(e){
						e.preventDefault();
						self.next();
					});
				}
				else {
					debug.warn('nextButton not specified');
				}

				// Setup select buttons
				if (s.selectButton && $(s.selectButton).length) {
					debug.log('creating selectButton');

					self.$selectButton = $(s.selectButton);

					// If constructor is provided, append to selectButton element
					if (s.selectHTML) {
						if (typeof s.selectHTML == 'string') {
							debug.log('adding selectHTML');
							// Check for valid HTML
							if (validateHTML(s.selectHTML, 'selectHTML')) {
								self.$selectButton = $(
									// Place as many buttons as slides
									//repeatString(s.selectHTML, self.numSlides)
									repeatElement(s.selectHTML, self.numSlides)
									// and append to only the first matched element
								).appendTo(self.$selectButton.eq(0));
								if (s.selectHTMLNumbers) {
									// Get type of element introduced by selectHTML, just in case there are other elements inside selectHTML
									var tagName = $(s.selectHTML).prop('tagName');
									$(s.selectButton).children(tagName).each(function(index) {
										$(this).append('<span>' + (index+1) + '</span>');
									});
								}
							}
						}
						else if (typeof s.selectHTML == 'function') {
							var selectHTMLStr = '';
							// Call the constructor for each slide, passing index
							self.$slides.each(function(i) {
								selectHTMLStr += s.selectHTML.call(this, i);
							});
							self.$selectButton = $(selectHTMLStr).appendTo(self.$selectButton);
						}
						// Error conditions
						else {
							if (typeof s.selectHTML == 'string' && s.selectHTML.charAt(0) !== '<') {
								alert('selectHTML must specify a valid HTML string');
							}
						}
					}

					self.$selectButton.each(function(i) {
						$(this).click(function(e) {
							e.preventDefault();
							self.select(i);
						});
					});
				}
				else {
					debug.warn('selectButton not specified');
				}
				debug.groupEnd();
			}
			else {
				if (s.verbose) {
					debug.group('not creating controls');
					if (self.numSlides === 0)
						debug.warn('only 1 slide, controls not created');
					else
						debug.log('useControls set to false');
					debug.groupEnd();
				}
			}
		};

		// Sets up the slides for specific mode of transition
		var animationSetup = function() {
			debug.log('as: animationSetup()');
			if (s.animate) {
				debug.group('setting up slider elements...')
				debug.log('mode = ' + s.mode);
				// Mode select
				switch (s.mode.charAt(0)) {
					case 'c': // custom
						s.customSetup.call(self, self.$slider);
						break;
					case 'f': // fade
						self.$slider.css({
							'display': 'block',
							'position': 'relative'
						}).append('<div style="clear: both"/>');
						self.$slides.css({
							'display': 'block',
							'float': 'left',
							'width': '100%',
							'margin-left': 0,
							'margin-right': '-100%',
							'opacity': 0,
							'visibility': 'hidden'
						}).eq(self.currentSlide).css({'opacity': 1, 'visibility': 'visible'});
						break;
					case 's': // slide
					default: // If input is not understood, slide is default
						self.$slider.css({
							'display': 'block',
							'position': 'relative',
							'white-space': 'nowrap',
							'font-size': 0, // removes spaces on inline elements
							'overflow': 'hidden'
						});
						self.$slides.css({
							'display': 'inline-block',
							'white-space': 'normal',
							'width': '100%',
							'font-size': '16px' // reset font size
						});
						break;
					// Room for future transition types?
				}
				debug.groupEnd();
			}
		};

		// Handles transitions between slides
		var transition = function(fromSlide, toSlide) {
			debug.log('as: transition(%d, %d)', fromSlide, toSlide);

			if (s.animate && !self.animating) {
				debug.group('animating...');
				self.animating = true;
				// Mode select
				switch (s.mode.charAt(0)) {
					case 'c': // custom
						debug.log('custom transition');
						s.customTransition.call(self, self.$slider, fromSlide, toSlide, s.transitionTime, s.easing, self.doneAnimating);
						break;
					case 'f': // fade
						debug.log('animation will fade');
						self.animate.fade(fromSlide, toSlide, s.transitionTime, s.easing, self.doneAnimating);
						break;
					case 's': // slide
					default: // If input is not understood, slide is default
						debug.log('animation will slide');
						self.animate.slide(fromSlide, toSlide, s.transitionTime, s.easing, self.doneAnimating);
						break;
					// Room for future transition types?
				}
				debug.groupEnd();
			}
			else if (self.animating) {
				debug.warn('as: animation is already ocurring!');
			}
		};

		// Handles changing slide classes, initiates transition, and callbacks
		var goToSlide = function(slide) {
			debug.log('as: goToSlide(%d)', slide);
			// Only go if not currently animating
			if (!self.animating) {
				var fromSlide = self.currentSlide,
					toSlide = slide;

				debug.group('transitioning slides...');

				// Only go if slide requested is different than the current one
				if (toSlide !== fromSlide) {

					debug.log('going from %d to %d', fromSlide, toSlide);

					// Trigger on will slide callback
					onWillSlideCallback(fromSlide, toSlide);

					// transition
					transition(fromSlide, toSlide);

					// Set classes on target slide, prev, next, & update select buttons
					self.setCurrentSlideAndSelect(toSlide);

					// autoPlay
					if (s.autoPlay && self.playing) {
						self.setTimer();
					}

					// Trigger on did slide callback
					onDidSlideCallback(fromSlide, toSlide);
				}
				else {
					debug.warn('already on slide %d', toSlide);
				}

				debug.groupEnd();
			}
			else {
				debug.warn('as: cannot go while animation is ocurring!')
			}
		};

		// Initialize
		return init();

	};

	// Add plugin to jQuery
	$.fn[pluginName] = function(options) {
		if (options && options.hasOwnProperty('verbose') && options.verbose) console.log('as: %s plugin invoked with -verbose', pluginName);

		return this.each(function() {
			var ss = $(this).data(pluginName);

			// Initialize while preventing multiples
			if (!ss)
				(new AdaptaSlider(this, options));
			else
				ss.update();
		});

		//return this;
	};

	// Global utility functions

	// Return integer modulo
	function modInt(num, denom) {
		//if(num > 0) return Math.floor(num % denom);
		//else return Math.ceil(num % denom);
		return ((num % denom) + denom) % denom;
	}

	// Return repeated string
	function repeatString(str, count) {
		return new Array(count+1).join(str);
	}

	// Return repeated jQuery element
	function repeatElement(str, count) {
		var $el;
		// This method avoids nesting if string is an unclosed HTML tag
		for (var i = 0; i < count; i++) {
			if (!$el) $el = $(str);
			else $el = $el.add(' '+str);
		}
		return $el;
	}

	// Validate HTML string
	function validateHTML(str, name) {
		var valid = false;
		try { $(str)[0]; valid = true; }
		catch (e) {
			alert('Your "'+name+'" HTML string did not pass validation:\n' + str + '\n' + e);
			console.error('as: Your "%s" HTML string did not pass validation: "%s"', name, str);
		}
		return valid;
	}

})(jQuery);
